src/bsoncxx/types/value.cpp

```diff
@@ -56,35 +74,52 @@ value& value::operator=(const value& rhs) {
 
 value::value(value&& rhs) noexcept {
     *this = std::move(rhs);
+    rhs._type = stdx::nullopt;
 }
 
 value& value::operator=(value&& rhs) noexcept {
-    _type = rhs._type;
+    if(_type && !rhs._type) {
```

Comment:
The author of [this stack overflow post](http://stackoverflow.com/questions/9322174/move-assignment-operator-and-if-this-rhs) feels that a check here isn't actually necessary.  The reasoning is that either rhs is a temporary, so it should be a unique reference, or rhs is some other object that the user wants to make look like a temporary.  In the second case, it's a user error if user code causes a self move assign, so we should consider throwing an exception here.